renderer/GLSLProgram.cpp

```diff
@@ -0,0 +1,334 @@
+/*****************************************************************************
+The Dark Mod GPL Source Code
+
+This file is part of the The Dark Mod Source Code, originally based
+on the Doom 3 GPL Source Code as published in 2011.
+
+The Dark Mod Source Code is free software: you can redistribute it
+and/or modify it under the terms of the GNU General Public License as
+published by the Free Software Foundation, either version 3 of the License,
+or (at your option) any later version. For details, see LICENSE.TXT.
+
+Project: The Dark Mod (http://www.thedarkmod.com/)
+
+******************************************************************************/
+
+#include "precompiled.h"
+#include "GLSLProgram.h"
+#include <regex>
+
+GLuint GLSLProgram::currentProgram = 0;
+
+GLSLProgram * GLSLProgram::Load( const char *vertexSourceFile, const char *fragmentSourceFile, const char *geometrySourceFile ) {
+	return Load( idDict(), vertexSourceFile, fragmentSourceFile, geometrySourceFile );	
+}
+
+GLSLProgram * GLSLProgram::Load( const idDict &defines, const char *vertexSourceFile, const char *fragmentSourceFile, const char *geometrySourceFile ) {
+	GLSLProgramLoader loader;
+	loader.AddVertexShader( vertexSourceFile, defines );
+	loader.AddFragmentShader( fragmentSourceFile, defines );
+	if( geometrySourceFile != nullptr ) {
+		loader.AddGeometryShader( geometrySourceFile, defines );
+	}
+	return loader.LinkProgram();
+}
+
+GLSLProgram::GLSLProgram( GLuint program ) : program( program ) {}
+
+GLSLProgram::~GLSLProgram() {
+	qglDeleteProgram( program );
+}
+
+void GLSLProgram::Activate() {
+	if( currentProgram != program ) {
+		qglUseProgram( program );
+		currentProgram = program;
+	}
+}
+
+void GLSLProgram::Deactivate() {
+	qglUseProgram( 0 );
+	currentProgram = 0;
+}
+
+void GLSLProgram::AddUniformAlias( int alias, const char *uniformName ) {
+	int location = qglGetUniformLocation( program, uniformName );
+	if( location == -1 ) {
+		common->Warning( "Did not find active uniform: %s\n", uniformName );
```

Comment:
This should **not** be a warning.

It is very frequent that uniform is declared in the program but not used. In such cases most OpenGL implementations remove uniform entirely, and this warning will be raised, despite the fact that uniform is declared. This will cause a lot of confusion and noise.

I'm afraid we have to silently map this `alias` into `-1` location.

Reply:
That depends strongly on the use case. My expectation is that at least for the core shaders we only pass in information that the shader actually needs. So if the uniform isn't actually used in the shader, that could hint at a misspelling or other mistakes, and so the warning could be quite helpful.

If we treat shaders more generically than that, then yes, the warning might be counter-productive.